Change InteracitonRegion to hold an IntRect instead of a Region

megangardner opened the pull request:
#### f51974aa4007bbd070c94d507d472bddd1e95745
<pre>
Change InteracitonRegion to hold an IntRect instead of a Region
<a href="https://bugs.webkit.org/show_bug.cgi?id=254456">https://bugs.webkit.org/show_bug.cgi?id=254456</a>
rdar://107216144

Reviewed by Tim Horton.

InteractionRegion currently has a Region as the way that it holds
the data about the interaction. However, Region is a large
data structure, and we only ever put a single Rect into it,
so to save space and performance, we should switch to just using a
Rect instead of a Region.

* LayoutTests/interaction-region/aria-roles-expected.txt:
* LayoutTests/interaction-region/click-handler-dynamically-added-expected.txt:
* LayoutTests/interaction-region/click-handler-expected.txt:
* LayoutTests/interaction-region/click-handler-in-shadowed-layer-expected.txt:
* LayoutTests/interaction-region/consolidated-nested-regions-expected.txt:
* LayoutTests/interaction-region/event-region-overflow-expected.txt:
* LayoutTests/interaction-region/full-page-overlay-expected.txt:
* LayoutTests/interaction-region/icon-inside-button-single-region-expected.txt:
* LayoutTests/interaction-region/inline-link-dark-background-expected.txt:
* LayoutTests/interaction-region/inline-link-expected.txt:
* LayoutTests/interaction-region/inline-link-in-composited-iframe-expected.txt:
* LayoutTests/interaction-region/inline-link-in-layer-expected.txt:
* LayoutTests/interaction-region/inline-link-in-non-composited-iframe-expected.txt:
* LayoutTests/interaction-region/inline-link-with-pointer-events-none-content-expected.txt:
* LayoutTests/interaction-region/input-type-file-region-expected.txt:
* LayoutTests/interaction-region/input-type-range-region-expected.txt:
* LayoutTests/interaction-region/nested-composited-text-painter-expected.txt:
* LayoutTests/interaction-region/overlap-expected.txt:
* LayoutTests/interaction-region/overlay-expected.txt:
* LayoutTests/interaction-region/paused-video-regions-expected.txt:
* LayoutTests/interaction-region/pseudo-element-expected.txt:
* LayoutTests/interaction-region/split-inline-link-expected.txt:
* LayoutTests/interaction-region/wrapped-inline-link-expected.txt:
* Source/WebCore/page/DebugPageOverlays.cpp:
(WebCore::pathsForRect):
(WebCore::InteractionRegionOverlay::activeRegion const):
(WebCore::InteractionRegionOverlay::drawRect):
(WebCore::pathsForRegion): Deleted.
* Source/WebCore/page/InteractionRegion.cpp:
(WebCore::interactionRegionForRenderedRegion):
(WebCore::operator&lt;&lt;):
* Source/WebCore/page/InteractionRegion.h:
(WebCore::operator==):
* Source/WebCore/rendering/EventRegion.cpp:
(WebCore::EventRegionContext::uniteInteractionRegions):
(WebCore::EventRegionContext::copyInteractionRegionsToEventRegion):
(WebCore::EventRegion::translate):
(WebCore::EventRegion::appendInteractionRegions):
(WebCore::EventRegion::uniteInteractionRegions): Deleted.
* Source/WebCore/rendering/EventRegion.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::paintDebugOverlays):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeInteractionRegionLayers.mm:
(WebKit::updateLayersForInteractionRegions):

Canonical link: <a href="https://commits.webkit.org/262242@main">https://commits.webkit.org/262242@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ee145ef460c279fb119d138945f6e1e5eb8cddcc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/982 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/1010 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/1047 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/1483 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/872 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/971 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/1062 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1095 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/1079 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/991 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/927 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/923 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/1383 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/971 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/919 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/911 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/898 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/947 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/1959 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/939 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/887 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/885 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/928 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/238 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/954 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

etiennesegonzac commented on Source/WebCore/rendering/EventRegion.cpp:
```diff
@@ -184,9 +184,21 @@ bool EventRegionContext::shouldConsolidateInteractionRegion(IntRect bounds, Rend
     return false;
 }
 
+/*void EventRegionContext::guardSmallIntetactionRegions()
```
small typo ("Intetaction"), even this will probably be part of the next patch

etiennesegonzac reviewed (commented):
Jinx :/ 
Had to use Region features to fix https://github.com/WebKit/WebKit/pull/11938

It might still be worth it to:
* only use a Region for `HashMap<ElementIdentifier, Region> m_discoveredRegionsByElement`
* handle the case where `regionInLayerCoordinates.subtract` creates multiple rects by appending multiple Rect-based-InteractionRegions to `m_interactionRegions` (we have a test for that case already)

(small typo in the commit message too, "Interaciton")

etiennesegonzac commented on Source/WebCore/rendering/EventRegion.cpp:
```diff
@@ -197,7 +210,8 @@ bool EventRegionContext::shouldConsolidateInteractionRegion(IntRect bounds, Rend
 void EventRegionContext::copyInteractionRegionsToEventRegion()
 {
-    m_eventRegion.uniteInteractionRegions(m_interactionRegions);
+    
```
nit: the extra blank line seems odd

etiennesegonzac commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeInteractionRegionLayers.mm:
```diff
@@ -116,40 +116,41 @@ void updateLayersForInteractionRegions(CALayer *layer, RemoteLayerTreeHost& host
+            if (!foundInPosition)
+                [layer insertSublayer:occlusionLayer.get() atIndex:insertionPoint];
 
+            [layer insertSublayer:occlusionLayer.get() atIndex: insertionPoint];
```
rebase issue?
we should only insert the occlusion layer in the `if (!foundInPosition)` above.

etiennesegonzac reviewed (commented):
This is great! And the way it cleans up the test expectations is really the cherry on top :)

r=me with the extra `insertSublayer` call removed, I'll let @hortont424 do the official review

hortont424 reviewed (approved):
please fix the typo in the title

megangardner committed:
Change InteracitonRegion to hold an IntRect instead of a Region
https://bugs.webkit.org/show_bug.cgi?id=254456
rdar://107216144

Reviewed by Tim Horton.

InteractionRegion currently has a Region as the way that it holds
the data about the interaction. However, Region is a large
data structure, and we only ever put a single Rect into it,
so to save space and performance, we should switch to just using a
Rect instead of a Region.

* LayoutTests/interaction-region/aria-roles-expected.txt:
* LayoutTests/interaction-region/click-handler-dynamically-added-expected.txt:
* LayoutTests/interaction-region/click-handler-expected.txt:
* LayoutTests/interaction-region/click-handler-in-shadowed-layer-expected.txt:
* LayoutTests/interaction-region/consolidated-nested-regions-expected.txt:
* LayoutTests/interaction-region/event-region-overflow-expected.txt:
* LayoutTests/interaction-region/full-page-overlay-expected.txt:
* LayoutTests/interaction-region/icon-inside-button-single-region-expected.txt:
* LayoutTests/interaction-region/inline-link-dark-background-expected.txt:
* LayoutTests/interaction-region/inline-link-expected.txt:
* LayoutTests/interaction-region/inline-link-in-composited-iframe-expected.txt:
* LayoutTests/interaction-region/inline-link-in-layer-expected.txt:
* LayoutTests/interaction-region/inline-link-in-non-composited-iframe-expected.txt:
* LayoutTests/interaction-region/inline-link-with-pointer-events-none-content-expected.txt:
* LayoutTests/interaction-region/input-type-file-region-expected.txt:
* LayoutTests/interaction-region/input-type-range-region-expected.txt:
* LayoutTests/interaction-region/nested-composited-text-painter-expected.txt:
* LayoutTests/interaction-region/overlap-expected.txt:
* LayoutTests/interaction-region/overlay-expected.txt:
* LayoutTests/interaction-region/paused-video-regions-expected.txt:
* LayoutTests/interaction-region/pseudo-element-expected.txt:
* LayoutTests/interaction-region/split-inline-link-expected.txt:
* LayoutTests/interaction-region/wrapped-inline-link-expected.txt:
* Source/WebCore/page/DebugPageOverlays.cpp:
(WebCore::pathsForRect):
(WebCore::InteractionRegionOverlay::activeRegion const):
(WebCore::InteractionRegionOverlay::drawRect):
(WebCore::pathsForRegion): Deleted.
* Source/WebCore/page/InteractionRegion.cpp:
(WebCore::interactionRegionForRenderedRegion):
(WebCore::operator<<):
* Source/WebCore/page/InteractionRegion.h:
(WebCore::operator==):
* Source/WebCore/rendering/EventRegion.cpp:
(WebCore::EventRegionContext::uniteInteractionRegions):
(WebCore::EventRegionContext::copyInteractionRegionsToEventRegion):
(WebCore::EventRegion::translate):
(WebCore::EventRegion::appendInteractionRegions):
(WebCore::EventRegion::uniteInteractionRegions): Deleted.
* Source/WebCore/rendering/EventRegion.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::paintDebugOverlays):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeInteractionRegionLayers.mm:
(WebKit::updateLayersForInteractionRegions):

Canonical link: https://commits.webkit.org/262242@main
